megatron/data/dataset_utils.py

```diff
@@ -231,24 +247,40 @@ def create_masked_lm_predictions(tokens,
     if masked_lm_prob == 0:
         return (output_tokens, masked_lm_positions,
                 masked_lm_labels, token_boundary)
+    if prefix_lm:
+        # Adjust probabilities so that the mean is centered at the
+        # correct position.
+        # If we do not do this, the mean is at
+        # `len(tokens) * masked_lm_prob / 2`.
+        masked_lm_prob *= 2
 
     num_to_predict = min(max_predictions_per_seq,
                          max(1, int(round(len(tokens) * masked_lm_prob))))
 
     ngrams = np.arange(1, max_ngrams + 1, dtype=np.int64)
-    if not geometric_dist:
+    if sampling_style is SamplingStyle.POISSON:
         # Note(mingdachen):
         # By default, we set the probilities to favor shorter ngram sequences.
         pvals = 1. / np.arange(1, max_ngrams + 1)
         pvals /= pvals.sum(keepdims=True)
         if favor_longer_ngram:
             pvals = pvals[::-1]
+    elif sampling_style is SamplingStyle.NORMAL:
+        normal_mean = (max_ngrams + 1) / 2
 
     ngram_indexes = []
     for idx in range(len(cand_indexes)):
         ngram_index = []
         for n in ngrams:
+            if prefix_lm:
+                last_cand_index_index = min(idx + n - 1, len(cand_indexes) - 1)
+                if cand_indexes[last_cand_index_index][-1] < len(tokens) - 1:
+                    continue
```

Comment:
It seems that lines 276-278 only run for `n=1`?
This snippet could use some more documentation.

Reply:
Correct, we [avoid redundant compute here](https://github.com/NVIDIA/Megatron-LM/pull/268/files/324d70d2518fe183c27e36ed550a95955eaa077c#diff-e1d14be32f4489a01cb8d571804fbba003f7f90715ef3cb3a27d9099e0245d6fR281-R282).
The snippet selects only those index sets for which the final index is at the end of the sequence. I will add this comment to the code as well.

Reply:
Sorry, I misunderstood you. No, this does not run only for `n = 1`. Due to the `continue`, we skip a couple `n` until we arrive at the first one for which the final index is at the end of the sequence. Then, we break out of the loop. So the `ngram_index.append` call only happens once for each inner `for`-loop, but not necessarily for `n = 0` or `n = 1`.

Reply:
This is now removed and we compute `cand_indexes` independently for the PrefixLM task.